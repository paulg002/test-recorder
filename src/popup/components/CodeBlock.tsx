import * as React from 'react';

export interface CodeBlockProps {
  index: number,
  text: string,
  dragStatus: string,
  destroyBlock: (i: number) => void,
  onDragStart: (e: React.DragEvent, i: number) => void,
  onDragOver: (e: React.DragEvent, i: number) => void,
  onDragEnd: () => void,
  onDrop: (e: React.DragEvent, i: number) => void,
}

export default ({
  index,
  text,
  destroyBlock,
  onDragStart,
  onDragOver,
  onDragEnd,
  onDrop,
  dragStatus,
}: CodeBlockProps) => {
  const i = text.startsWith('cy.url()')
    ? text.lastIndexOf('\'', text.length - 4)
    : text.indexOf('\'');
  const j = text.lastIndexOf('\'') + 1;
  const preSelector = text.slice(0, i);
  const selector = text.slice(i, j);
  const postSelector = text.slice(j);
  return (
    <li
      className={dragStatus}
      draggable
      onDragStart={e => onDragStart(e, index)}
      onDragEnd={onDragEnd}
      onDragOver={e => onDragOver(e, index)}
      onDrop={e => onDrop(e, index)}
    >
      <span>
        {preSelector}
        <mark className="selector">
          {selector}
        </mark>
        {postSelector}
      </span>
      <button type="button" className="delete" onClick={() => destroyBlock(index)}>x</button>
    </li>
  );
};
