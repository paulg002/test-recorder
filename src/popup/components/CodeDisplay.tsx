import * as React from 'react';
import CodeBlock from './CodeBlock';
import type { Block } from '../../types';

export interface CodeDisplayProps {
  codeBlocks: Block[],
  destroyBlock: (index: number) => void,
  moveBlock: (dragIdx: number, dropIdx: number) => void,
}

export default ({ codeBlocks, destroyBlock, moveBlock }: CodeDisplayProps) => {
  const [dragOriginIdx, setDragOriginIdx] = React.useState<number>(-1);
  const [draggedIdx, setDraggedIdx] = React.useState<number>(-1);
  const [modifiedCodeBlocks, setModifiedCodeBlocks] = React.useState<Block[]>([]);

  React.useEffect(() => {
    setModifiedCodeBlocks([...codeBlocks]);
  }, [codeBlocks]);

  const swapBlocks = (i: number, j: number) => {
    const temp = [...modifiedCodeBlocks];
    const holder = temp[i];
    temp[i] = temp[j];
    temp[j] = holder;
    setModifiedCodeBlocks(temp);
  };

  const onDragStart = (e: React.DragEvent, i: number) => {
    setDraggedIdx(i);
    setDragOriginIdx(i);
  };

  const onDragOver = (e: React.DragEvent, i: number) => {
    e.preventDefault();
    if (draggedIdx !== i) {
      swapBlocks(draggedIdx, i);
      setDraggedIdx(i);
    }
  };

  const onDragEnd = () => {
    setDraggedIdx(-1);
    setModifiedCodeBlocks([...codeBlocks]);
  };

  const onDrop = (e: React.DragEvent, i: number) => {
    e.preventDefault();
    if (dragOriginIdx !== i) moveBlock(dragOriginIdx, i);
    setDragOriginIdx(-1);
  };

  const blocks = modifiedCodeBlocks.map((block, idx) => (
    <CodeBlock
      key={block.id}
      dragStatus={idx === draggedIdx ? 'drag-origin' : 'code-block'}
      text={block.value}
      index={idx}
      destroyBlock={destroyBlock}
      onDragStart={onDragStart}
      onDragOver={onDragOver}
      onDragEnd={onDragEnd}
      onDrop={onDrop}
    />
  ));

  return (
    <ul id="code-display">
      {blocks}
    </ul>
  );
};
