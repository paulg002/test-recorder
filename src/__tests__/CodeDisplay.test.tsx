import * as React from 'react';
import { shallow } from 'enzyme';
import CodeDisplay, { CodeDisplayProps } from '../popup/components/CodeDisplay';
import CodeBlock from '../popup/components/CodeBlock';
import '../setupTests';

describe('CodeDisplay', () => {
  let wrapper;
  const props: CodeDisplayProps = {
    codeBlocks: [
      {
        id: 'id1',
        value: 'value1',
      },
      {
        id: 'id2',
        value: 'value2',
      },
      {
        id: 'id3',
        value: 'value3',
      },
    ],
    destroyBlock: jest.fn(),
    moveBlock: jest.fn(),
  };
  beforeAll(() => {
    wrapper = shallow(<CodeDisplay {...props} />);
  });

  it('Should render an array of CodeBlocks', () => {
    expect(wrapper).toMatchSnapshot();
    // expect(wrapper.html()).toBe(false);
    // expect(wrapper.exists(CodeBlock)).toBe(true);
    // const blocks = wrapper.find(CodeBlock);
    // expect(blocks).toHaveLength(3);
    // for (let i = 0; i !== props.codeBlocks.length; i += 1) {
    //   expect(blocks[i].key()).toBe(props.codeBlocks[i].id);
    // }
  })
});
