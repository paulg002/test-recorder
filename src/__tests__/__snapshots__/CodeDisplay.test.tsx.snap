// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodeDisplay Should render an array of CodeBlocks 1`] = `
<ul
  id="code-display"
/>
`;
